add pm2.5 DA for rrfs-smoke

hongli-wang opened the pull request:
Add PM2.5 DA capability for RRFS-SMOKE.

hongli-wang committed:
1. add copy ges for rrfs-cmaq
2. add pm2.5 DA for rrfs-smoke

hongli-wang committed:
Refine PM2.5 DA for RRFS-SMOKE

hongli-wang committed:
Refine use of PM2.5 for RRFS-SD;

CoryMartin-NOAA commented:
@hongli-wang I'm going to be starting on the review for this (sorry it has taken so long). Can you merge in 'develop' to your branch and we will get started. Do you know of anyone who knows GSI and AQ well enough to be a second reviewer besides myself?

CoryMartin-NOAA commented on src/gsi/gsi_rfv3io_mod.f90:
```diff
@@ -1101,6 +1116,31 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
+              write(6,*)'the chemvarname ',jtracer,vartem,' is found '
+           else
+              write(6,*)'the chemvarname ',vartem,' is not in aeronames_smoke_fv3, !!!!!!!!!!'
+              !call flush(6)
```
is this a critical error or is it okay for it to warn and continue?

hongli-wang replied:
> is this a critical error or is it okay for it to warn and continue?

Let's see if I can improve this. For my application, this is not a critical error. For example, the pm2.5 is one of guess field, but it is not a model prognostic variable.

hongli-wang commented:
> @hongli-wang I'm going to be starting on the review for this (sorry it has taken so long). Can you merge in 'develop' to your branch and we will get started. Do you know of anyone who knows GSI and AQ well enough to be a second reviewer besides myself?

Do you mean rebase my feature/RRFS-CMAQ branch to the EMC GSI develop branch? What I see now is that my branch doesn't have conflicts and can merge to EMC GSI develop branch. 

Thanks for reviewing my GSI code. The others I can recommend are Ting Lei, Ming Hu and Mariusz Pagowski. Ting reviewed my previous update on RRFS-CMAQ DA, so if he would like to review the code this time would be best.

CoryMartin-NOAA commented:
@TingLei-daprediction would you have time to be able to provide a review of this PR alongside my review?

CoryMartin-NOAA commented:
@hongli-wang yes I see there are no conflicts so it should be fine (at least for now). I will proceed with starting my review.

TingLei-daprediction commented:
@CoryMartin-NOAA Sure. I will begin my review as soon as possible.

CoryMartin-NOAA commented on src/gsi/gsi_rfv3io_mod.f90:
```diff
@@ -1101,6 +1116,31 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
+           endif
+        enddo
+        ntracersmokeio3d=jtracer+1
+        fv3lam_io_tracersmokevars3d_nouv(jtracer+1)="pm2_5"
```
probably a dumb question, but what does `nouv` mean on all of these? is it 'no wind' or 'no ultraviolet'? or something else?

TingLei-NOAA replied:
It means "no wind".

hongli-wang replied:
> It means "no wind".

Hi Ting, 

Thanks for your time and review. I will address them in my latter commits.

Hongli

CoryMartin-NOAA commented on src/gsi/gsi_rfv3io_mod.f90:
```diff
@@ -1324,12 +1405,36 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
+           call GSI_BundleGetPointer ( gsibundle_fv3lam_tracersmoke_nouv,'dust',ges_dust,istatus );ier=ier+istatus
+           call GSI_BundleGetPointer ( gsibundle_fv3lam_tracersmoke_nouv,'pm2_5',ges_pm2_5,istatus );ier=ier+istatus
+           ! if (ier/=0) call die(trim(myname),'cannot get pointers for fv3    !  chem-fields, ier =',ier)
+           if (ier/=0) write(6,*),"cannot get pointers for fv3 smoke" 
```
shouldn't it crash if it cannot get the pointer?

hongli-wang replied:
> shouldn't it crash if it cannot get the pointer?

I will call the die subroutine. Thanks.

CoryMartin-NOAA commented on src/gsi/intpm2_5.f90:
```diff
@@ -307,6 +308,127 @@ subroutine intpm2_5_(pm2_5head,rval,sval)
 
+!
+   if (laeroana_fv3smoke) then
+       !pm2_5ptr => pm2_5head
```
remove this commented out line if it is unneeded

hongli-wang replied:
> remove this commented out line if it is unneeded

Yes, will do.

CoryMartin-NOAA commented on src/gsi/intpm2_5.f90:
```diff
@@ -307,6 +308,127 @@ subroutine intpm2_5_(pm2_5head,rval,sval)
+          val= val*pm2_5ptr%pm25wc(iaero)
+
+          nullify(spm2_5)
+!
```
empty comment?

hongli-wang replied:
I will delete it.

CoryMartin-NOAA commented on src/gsi/intpm2_5.f90:
```diff
@@ -307,6 +308,127 @@ subroutine intpm2_5_(pm2_5head,rval,sval)
+                if( ladtest_obs ) then
+                   grad = val
+                else
+                   grad     = val*pm2_5ptr%raterr2*pm2_5ptr%err2
```
why all these extra spaces left of the =?

CoryMartin-NOAA commented on src/gsi/intpm2_5.f90:
```diff
@@ -307,6 +308,127 @@ subroutine intpm2_5_(pm2_5head,rval,sval)
+
+!       adjoint
+             !aeroname='smoke'
+          do iaero=1, naero_smoke_fv3   
```
indentation is weird here

hongli-wang replied:
I will re-format the lines.

CoryMartin-NOAA commented on src/gsi/intpm2_5.f90:
```diff
@@ -307,6 +308,127 @@ subroutine intpm2_5_(pm2_5head,rval,sval)
+             rpm2_5(j7)=rpm2_5(j7)+w7*grad*pm2_5ptr%pm25wc(iaero)
+             rpm2_5(j8)=rpm2_5(j8)+w8*grad*pm2_5ptr%pm25wc(iaero)
+             nullify(rpm2_5)
+          end do
```
Make sure indentation is consistent/correct throughout this routine.

hongli-wang replied:
Sure.  I will re-format the lines.

CoryMartin-NOAA commented on src/gsi/setuppm2_5.f90:
```diff
@@ -669,17 +736,43 @@ subroutine setuppm2_5(obsLL,odiagLL,lunin,mype,nreal,nobs,isis,is,conv_diagsave)
+                 innov = 0.0_r_kind 
+              end if
+              if (tv_ges-273.15_r_kind < 0.0_r_kind) then
+                 innov = 0.0_r_kind
```
no innovations below freezing?

hongli-wang replied:
RRFS-SD only has one smoke and one dust tracer. So that we only can assimilate the smoke/dust dominated PM2.5. However, this is a challenge to identify which station is smoke/dust affected from the value itself. In addition to ue of smoke/dust mask from satellite in obs preprocess, here a physical constraint of temperature is used to exclude non-smoke affected  PM2.5 obs since the tempature should be greater than 0.0 where fire takes place. We may adjust this in future after more tests.

CoryMartin-NOAA commented on src/gsi/setuppm2_5.f90:
```diff
@@ -669,17 +736,43 @@ subroutine setuppm2_5(obsLL,odiagLL,lunin,mype,nreal,nobs,isis,is,conv_diagsave)
+                mype,nfldsig)
+             call tintrp2a11(pm25wc(:,:,:,2,nfldsig),pm25wc_ges(2),dlat,dlon,dtime,hrdifsig,&
+                mype,nfldsig)
+              print*,"tv_ges= ",tv_ges,conc,pm2_5ges,pm25wc_ges(1:2)
```
remove this debug print

hongli-wang replied:
Will do. Thanks.

CoryMartin-NOAA commented on src/gsi/stppm2_5.f90:
```diff
@@ -255,6 +256,97 @@ subroutine stppm2_5(pm2_5head,rval,sval,out,sges,nstep)
+       pm2_5ptr => pm2_5Node_typecast(pm2_5head)
+
+       do while (associated(pm2_5ptr))
+          if(pm2_5ptr%luse)then
```
These if statements are confusing, where are the else/end ifs? Be consistent with indentation and perhaps add comments showing where the else/endif is.

hongli-wang replied:
Yes. I will do it.

CoryMartin-NOAA reviewed (commented):
A couple of comments here from the first round of review.

TingLei-NOAA commented on src/gsi/gsi_rfv3io_mod.f90:
```diff
@@ -915,15 +926,15 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
-    if (laeroana_fv3cmaq) then
+    if (laeroana_fv3cmaq.or.laeroana_fv3smoke) then
       if (.not.allocated(name_chemvars3d)) then
         allocate( name_chemvars3d(GSI_ChemGuess_Bundle(it)%n3d))
```
@hongli-wang  could we add a check in an appropriate place on if GSI_chemGuess_bundle(it)%n3d=naero_cmaq_fv3 +naero_smoke_fv3  (with possible additional numbers) ?   I think this will help users avoid possible problems in their setup

hongli-wang replied:
> @hongli-wang could we add a check in an appropriate place on if GSI_chemGuess_bundle(it)%n3d=naero_cmaq_fv3 +naero_smoke_fv3 (with possible additional numbers) ? I think this will help users avoid possible problems in their setup

I will think about this. Thanks.

TingLei-NOAA commented on src/gsi/gsi_rfv3io_mod.f90:
```diff
@@ -1866,8 +1971,8 @@ subroutine gsi_fv3ncdf2d_read(fv3filenamegin,it,ges_z,ges_t2m,ges_q2m)
 
-       deallocate (dim_id,sfc,sfc1,dim)
-       deallocate (sfc_fulldomain)
+       if(allocated(sfc1) .and. allocated(sfc))deallocate (dim_id,sfc,sfc1,dim)
```
Why do we need  first evaluating if sfc1 (and so on) has been allocated?   if "False", does that mean something wrong in prior running, which need to be corrected?

hongli-wang replied:
> Why do we need first evaluating if sfc1 (and so on) has been allocated? if "False", does that mean something wrong in prior running, which need to be corrected?

I can make here unchanged. The issue is that if we don't use the option of T2 and Q2 as guess from model, GSI will crash because it can't deallocate the arrays that have not been allocated. I think Ming will resolve this in his future PR.

hongli-wang replied:
I keep the allocated check before deallocate to keep GSI running when T2 and Q2 are not used as a guess variable from background files.

TingLei-NOAA commented on src/gsi/intpm2_5.f90:
```diff
@@ -307,6 +308,127 @@ subroutine intpm2_5_(pm2_5head,rval,sval)
+!naero_smoke_fv3
+          iaero=1
+          aeroname=aeronames_smoke_fv3(iaero) !'smoke'
+          call gsi_bundlegetpointer(sval,trim(aeroname),spm2_5,istatus)
```
@hongli-wang why don't define the pointers to corresponding sval component out of the loop over observations( loop 315)? Namely move those lines from line 333 out of this loop.

hongli-wang replied:
I understand. I keep it here for future to add more tracers in a do loop. spm2_5 is used here as a temporary array to count contribution from each tracer.

TingLei-NOAA replied:
@hongli-wang Got your point.  However, for the future case while several variables are to be accessed, we don't have to stick to a single point and, hence, use a loop of definition/nullyfy to access them.  We can just use , like, array of pointers to do that.   A loop of definition/nullyfy of a single pointer to be put in the loop over observations would seem, to me, even more inefficient.  We could define them outside of the observation loop and do the actual operations on them in a loop nested in the observation loop.

hongli-wang replied:
Agree. Let's make the code work correct first and then improve the efficiency.

TingLei-NOAA commented on src/gsi/intpm2_5.f90:
```diff
@@ -307,6 +308,127 @@ subroutine intpm2_5_(pm2_5head,rval,sval)
+               w7* spm2_5(j7)+w8* spm2_5(j8)
+          val= val*pm2_5ptr%pm25wc(iaero)
+
+          nullify(spm2_5)
```
it could be moved outside of the loop over observations.

hongli-wang replied:
> it could be moved outside of the loop over observations.

Same reason the the previous one. spm2_5 is used here as a temporary array to count contribution from each tracer. RRFS-SD may add more tracers in future.

TingLei-NOAA commented on src/gsi/setuppm2_5.f90:
```diff
@@ -232,6 +232,73 @@ subroutine setuppm2_5(obsLL,odiagLL,lunin,mype,nreal,nobs,isis,is,conv_diagsave)
+
+  endif 
+
   if (fv3_cmaq_regional .and. laeroana_fv3cmaq) then
```
@hongli-wang Could you explain what the two options (fv3_cmaq_regional  and laeroana_fv3cmaq) mean?

hongli-wang replied:
That means the model is regional FV3-CMAQ (fv3_cmaq_regional=.true.), and produce the analysis for the regional FV3-CMAQ (laeroana_fv3cmaq=.true.). I may add a description in the beginning of the subroutine.

TingLei-NOAA replied:
Thanks

TingLei-NOAA commented on src/gsi/stppm2_5.f90:
```diff
@@ -255,6 +256,97 @@ subroutine stppm2_5(pm2_5head,rval,sval,out,sges,nstep)
+          !!!
+          do iaero=1,naero_smoke_fv3
+             aeroname=aeronames_smoke_fv3(iaero)
+                call gsi_bundlegetpointer(sval,trim(aeroname),spm2_5,istatus)
```
The same comment as above, why don't do this outside of the loop (line 262).

hongli-wang replied:
As the reply to the previous comment, spm2_5 is a temporary array to hold contribution from each tracer. If only one tracer (or same variable to obs variable) it can be moved to outside of the loop, otherwise, it needs to keep here.

TingLei-NOAA commented:
@hongli-wang  Have you finished the existing regression tests for this PR? I know they are only available on hera now.

hongli-wang commented:
> @hongli-wang Have you finished the existing regression tests for this PR? I know they are only available on hera now.

Do you know where I can find guidance on the regression test? I haven't done that.

CoryMartin-NOAA commented:
@hongli-wang @TingLei-noaa , I/we (the GSI review team) will handle the regression testing, but I wanted to wait until it was closer to the 'final' version of the PR before submitting the tests.

hongli-wang reviewed (commented):
I am not sure why my replies are shown as comments.

hongli-wang commented:
> A couple of comments here from the first round of review.

Hi Corry,

Thanks for your comments. I will address them in my latter commit.

CoryMartin-NOAA commented:
No need to close this PR, we can keep it open for subsequent commits

hongli-wang commented:
> No need to close this PR, we can keep it open for subsequent commits

Understand. Thank you!

TingLei-NOAA commented:
> > @hongli-wang Have you finished the existing regression tests for this PR? I know they are only available on hera now.
> 
> Do you know where I can find guidance on the regression test? I haven't done that.
@RussTreadon-NOAA has a wiki on regression tests which is to be updated.  I will open an issue soon to document what is being done on hera with the link to former.

TingLei-NOAA commented:
@Cory Martin - NOAA Federal ***@***.***> When some developers
have no access to hera,  the reviewers take care of the regression tests is
essential .    However, for those with access to hera,  the regression
tests would be a necessary tool for their development , even before they
make their PR , not only as a criterion to pass but also as a helping tool
.   In fact, i would think it would be helpful if every PR could
offer their own regression test to verify their new functions to the
existing regression tests, though this seems not doable in the near
future.   So, when as you stated, the reviewing side would take care of the
regression tests, I would still open an issue with the link to GSI wiki on
how to do the regression tests and add my update on current practice on
hera to provide the options for GSI developers on hera.

______________________________

Ting Lei

Physical Scientist, Contractor with Lynker in support of

EMC/NCEP/NWS/NOAA

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

***@***.***

301-683-3624


On Tue, Dec 6, 2022 at 1:47 PM Cory Martin ***@***.***> wrote:

> @hongli-wang <https://github.com/hongli-wang> @TingLei-NOAA
> <https://github.com/TingLei-NOAA> , I/we (the GSI review team) will
> handle the regression testing, but I wanted to wait until it was closer to
> the 'final' version of the PR before submitting the tests.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/GSI/pull/482#issuecomment-1339820263>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APEFS7DPYTHIZHPLOSSYG4LWL6C5RANCNFSM6AAAAAAQ4ZE2IE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hu5970 commented:
> 

More developers can run regression test suite, the better for code development and review. We should update wiki page on how to run regression test as the steps may already be changed, at least they are not matching the steps I did.

hongli-wang commented:
> No need to close this PR, we can keep it open for subsequent commits

@CoryMartin-NOAA @TingLei-daprediction 

Hi Cory and Ting, 

I had pushed back my changes. Would you please review the codes again? 

Thanks again for your suggestions and comments! 


Hongli

CoryMartin-NOAA commented:
@hongli-wang thank you for addressing my comments!
@TingLei-NOAA please review again when you get a chance.

I'm not sure I'll be able to get to do the regression tests before I go on leave, but will do so in the new year. Once we know those pass + 2 thumbs up reviews, we can proceed with the merge.

hongli-wang commented:
> @hongli-wang thank you for addressing my comments! @TingLei-NOAA please review again when you get a chance.
> 
> I'm not sure I'll be able to get to do the regression tests before I go on leave, but will do so in the new year. Once we know those pass + 2 thumbs up reviews, we can proceed with the merge.

Totally understand since it is holiday and conference season. Please keep me posted. Thanks.

TingLei-NOAA commented:
@hongli-wang Did you push your changes during the PR reviewing to your branch feature/RRFS-CMAQ at  https://github.com/hongli-wang/GSI.git?  I just cloned it out and found the newest log is from Oct. 18,2022
^^
[Ting.Lei@hfe06 GSI]$ git remote -v
origin  https://github.com/hongli-wang/GSI.git (fetch)
origin  https://github.com/hongli-wang/GSI.git (push)
[Ting.Lei@hfe06 GSI]$ git pull
Already up to date.
[Ting.Lei@hfe06 GSI]$ git branch
* feature/RRFS-CMAQ
  master
[Ting.Lei@hfe06 GSI]$ git log
commit dfef936367734a8d4c406463a8a05c00c79bce57 (HEAD -> feature/RRFS-CMAQ, origin/feature/RRFS-CMAQ)
Author: hongli-wang <hongli.wang@noaa.gov>
VV
Did you directly commit your changes to the PR?   Your clarification will be helpful. 
About how to run regression tests on hera, I just opened an issue https://github.com/NOAA-EMC/GSI/issues/513. If you have access to hera, you could run regression tests .  Of course, I  am not sure if the current PR policy need the regression tests to be run by the reviewers . @CoryMartin-NOAA could clarify on that. 
In addition to those existing regression tests, I will run a few fv3-lam gsi tests using my case after I know how to get your updated PR.

hongli-wang commented:
> @hongli-wang Did you push your changes during the PR reviewing to your branch feature/RRFS-CMAQ at https://github.com/hongli-wang/GSI.git? I just cloned it out and found the newest log is from Oct. 18,2022 ^^ [Ting.Lei@hfe06 GSI]$ git remote -v origin https://github.com/hongli-wang/GSI.git (fetch) origin https://github.com/hongli-wang/GSI.git (push) [Ting.Lei@hfe06 GSI]$ git pull Already up to date. [Ting.Lei@hfe06 GSI]$ git branch
> 
> * feature/RRFS-CMAQ
>   master
>   [Ting.Lei@hfe06 GSI]$ git log
>   commit [dfef936](https://github.com/NOAA-EMC/GSI/commit/dfef936367734a8d4c406463a8a05c00c79bce57) (HEAD -> feature/RRFS-CMAQ, origin/feature/RRFS-CMAQ)
>   Author: hongli-wang [hongli.wang@noaa.gov](mailto:hongli.wang@noaa.gov)
>   VV
>   Did you directly commit your changes to the PR?   Your clarification will be helpful.
>   About how to run regression tests on hera, I just opened an issue [update on current GSI regression tests  #513](https://github.com/NOAA-EMC/GSI/issues/513). If you have access to hera, you could run regression tests .  Of course, I  am not sure if the current PR policy need the regression tests to be run by the reviewers . @CoryMartin-NOAA could clarify on that.
>   In addition to those existing regression tests, I will run a few fv3-lam gsi tests using my case after I know how to get your updated PR.

Hi Ting, 
It is correct. I used amend option to commit and forced to push the changes. So the log was not changed. 
Hongi

CoryMartin-NOAA commented:
@hongli-wang just an FYI, there is no need to do any of that anymore. We can just simply do a 'squash commit' when this is merged into the main branch.

hongli-wang commented:
> @hongli-wang just an FYI, there is no need to do any of that anymore. We can just simply do a 'squash commit' when this is merged into the main branch.

This is a good news. Glad to hear that!

TingLei-NOAA commented:
@hongli-wang  Thanks for the clarification about the log information.   
Based on our related discussion, I still think "call gsi_bundlegetpointer( .. " should not be done in a loop over observations (even for future multiple pointers cases) and the changes are needed from my point of view. 
Thanks.

hongli-wang commented:
> @hongli-wang Thanks for the clarification about the log information.
> Based on our related discussion, I still think "call gsi_bundlegetpointer( .. " should not be done in a loop over observations (even for future multiple pointers cases) and the changes are needed from my point of view.
> Thanks.

Thanks, Ting. I will look into this again. 

Hongli

hongli-wang committed:
Move getpointer out of obs do loop for rrfs_sd pm2.5 assimilation

hongli-wang commented:
> @hongli-wang Thanks for the clarification about the log information.
> Based on our related discussion, I still think "call gsi_bundlegetpointer( .. " should not be done in a loop over observations (even for future multiple pointers cases) and the changes are needed from my point of view.
> Thanks.

Hi Ting, 

Following your suggestion, I had modified the intpm2_5.f90 and stppm2_5.f90. Please review the changes at your convenience.

Thanks, 
Hongli

TingLei-daprediction commented:
@hongli-wang Thanks. I will finish my reviewing with a couple of fv3-lam tests as soon as possible. 
I just noticed that call of *getpointer* in the obs loop is in other existing code too like codes for pm10.  But we can leave them to the future to see any chances to update them.

hongli-wang commented:
> @hongli-wang Thanks. I will finish my reviewing with a couple of fv3-lam tests as soon as possible.
> I just noticed that call of _getpointer_ in the obs loop is in other existing code too like codes for pm10. But we can leave them to the future to see any chances to update them.

Regarding GETPOINTER: 
If an obs operator is a function of multiple state variables, it may need to define an array of pointers to hold the pointer for each state variables, otherwise need to define two new temporary pointers for each state variables. 

Please let me know when you see any issues. 

Thanks, 
Hongli

TingLei-daprediction commented:
@hongli-wang Yes, for example, see those treatments for dbz observation

TingLei-daprediction commented:
@hongli-wang the PR failed with my fv3-lam GSI test.  I found some piece in current emc-gsi is missing in your PR. 
like l_both_fv3sar_gfs_ens in gsimod.F90 and the mix ens option in association with this parameter. 
It is strange to me why those differences are not displayed in the "files changed"  section. 
Would you double check on this ?  (the branch I am testing locally is 
`[Ting.Lei@hfe07 gsi]$ git branch
* feature/RRFS-CMAQ
  master
[Ting.Lei@hfe07 gsi]$ git remote -v
origin  https://github.com/hongli-wang/GSI.git (fetch)
origin  https://github.com/hongli-wang/GSI.git (push)
`
Thanks

TingLei-daprediction commented:
@hongli-wang I am wondering if that is because your fork was forked from my fork, while my fork hadn't been updated for some time. Thuse the diff on github missed those changes (see https://stackoverflow.com/questions/55479729/github-pull-request-shows-wrong-diff).  Now, I have updated the develop branch of my fork. Please see if that help.

TingLei-daprediction commented:
And, now, all important function in my fork has been merged back to emc gsi and you could set your upstream to the emc-gsi if you haven't done that yet.

hongli-wang commented:
> @hongli-wang the PR failed with my fv3-lam GSI test. I found some piece in current emc-gsi is missing in your PR. like l_both_fv3sar_gfs_ens in gsimod.F90 and the mix ens option in association with this parameter. It is strange to me why those differences are not displayed in the "files changed" section. Would you double check on this ? (the branch I am testing locally is `[Ting.Lei@hfe07 gsi]$ git branch
> 
> * feature/RRFS-CMAQ
>   master
>   [Ting.Lei@hfe07 gsi]$ git remote -v
>   origin  https://github.com/hongli-wang/GSI.git (fetch)
>   origin  https://github.com/hongli-wang/GSI.git (push)
>   `
>   Thanks

Hi Ting, 

My GSI base is the EMC GSI develop branch on July 28 2022:  
Author: michael.lueken <Michael.Lueken@noaa.gov>
Date:   Thu Jul 28 17:05:42 2022 +0000

    GitHub Issue NOAA-EMC/GSI#444. Addition of EFSOI block to global_anavinfo.l127.txt.

If any RRs pushed back to the develop branch after the date, I will need to rebase to the latest develop branch before your PR tests. 

I will check. 

Thanks, 
Hongli

TingLei-NOAA commented:
@hongli-wang Yes.  then, I would suggest you merge/rebase the current emc gsi with your PR.

hongli-wang committed:
rebase to the latest EMC GSI develop branch

hongli-wang commented:
> @hongli-wang Yes. then, I would suggest you merge/rebase the current emc gsi with your PR.

Hi Ting, 

Just rebased to the latest EMC GSI develop branch: 

Author: RussTreadon-NOAA <26926959+RussTreadon-NOAA@users.noreply.github.com>
Date:   Fri Dec 16 11:39:26 2022 -0500

    Refactor global DA regression tests (#514)

Please let me know if you see any issues. 

thanks, 
Hongli

TingLei-daprediction commented:
@hongli-wang  Thanks for taking care of this!

TingLei-daprediction reviewed (dismissed):
This PR has passed my fv3-lam GSI test by producing the identical results as shown in the standard output.  All questions/concerns have been answered.  
Thanks to the developer's efforts for this PR.   Approved.

hongli-wang commented:
> This PR has passed my fv3-lam GSI test by producing the identical results as shown in the standard output. All questions/concerns have been answered.
> Thanks to the developer's efforts for this PR. Approved.

Hi Ting, 

Thanks for your review, having the code your fv3-lam tested, and approved this PR. 

Have a good Holiday! 

Hongli

CoryMartin-NOAA commented:
@hongli-wang I think the final thing we need to do here is modify the `fix` directory now that gerrit is no longer being used. Are there any fix file changes for your PR? We now have binary fix files staged on supported platforms (and on FTP) and only track ASCII fix files on the new gsi-fix GitHub repository.

hongli-wang commented:
> @hongli-wang I think the final thing we need to do here is modify the `fix` directory now that gerrit is no longer being used. Are there any fix file changes for your PR? We now have binary fix files staged on supported platforms (and on FTP) and only track ASCII fix files on the new gsi-fix GitHub repository.

@CoryMartin-NOAA It doesn't require to change any fix files for this PR right now.  Thanks.

CoryMartin-NOAA committed:
change submodule for fix to use github

CoryMartin-NOAA commented:
@hongli-wang we recently changed the GSI fix file approach, can you give me permission to push to your fork/branch or otherwise I will probably have to walk you through updating the fix submodule.

hongli-wang commented:
> @hongli-wang we recently changed the GSI fix file approach, can you give me permission to push to your fork/branch or otherwise I will probably have to walk you through updating the fix submodule.

@CoryMartin-NOAA Just invited you as a collaborator of my GSI fork.  I assume you will have the permission after you accept the invitation by default. Please let me know if you can't.

CoryMartin-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/RRFS-CMAQ

CoryMartin-NOAA committed:
update submodule to resolve change

CoryMartin-NOAA reviewed (dismissed):
Thank you, @hongli-wang, for all your hard work on this. Everything looks good to me, and a prior approval from @TingLei-daprediction / @TingLei-NOAA (including running the regression tests) means I think this is ready to go.

However, since I contributed to this PR by incorporating the new `fix/` submodule and merging in the latest changes from develop, I think I would like @ShunLiu-NOAA @hu5970 or @RussTreadon-NOAA to merge it in (and double check anything if they would like)

ShunLiu-NOAA commented:
> 

Cory, Hongli and Ting,

It appears that the regression test was completed in December, but the RRFS test case was not yet available in December. Could you re-run the RRFS regression test if you haven't do this already? Otherwise, it is ready to merge.

Thanks.

Shun

CoryMartin-NOAA commented:
@ShunLiu-NOAA good point, let me set that up as it has been a while since I've done the process myself.

RussTreadon-NOAA commented on src/gsi/gsi_rfv3io_mod.f90:
```diff
@@ -881,7 +891,8 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
     integer(i_kind):: inner_vars,numfields
     integer(i_kind):: ndynvario2d,ntracerio2d,ilev,jdynvar,jtracer
-    integer(r_kind):: iuv,ndynvario3d,ntracerio3d,ntracerchemio3d
+    integer(r_kind):: iuv,ndynvario3d,ntracerio3d
```
Why is this integer declared to be type `r_kind`?

CoryMartin-NOAA replied:
great catch! and that was already in develop...

hongli-wang replied:
I thought this has been resolved since in a previous version of my code for CMAQ it was fixed. Good to fix though It doesn't affect results.

RussTreadon-NOAA commented on src/gsi/intpm2_5.f90:
```diff
@@ -91,8 +92,8 @@ subroutine intpm2_5_(pm2_5head,rval,sval)
     real(r_kind) w1,w2,w3,w4,w5,w6,w7,w8
 ! real(r_kind) penalty
```
This line is commented out in both develop and this PR.  Is there any reason to keep this commented out line?  If not, please remove it.

hongli-wang replied:
The variable is not used, sure the line can be removed.

RussTreadon-NOAA commented on src/gsi/intpm2_5.f90:
```diff
@@ -226,7 +227,7 @@ subroutine intpm2_5_(pm2_5head,rval,sval)
           val=val *pm2_5ptr%pm25wc(imodes_cmaq_fv3(iaero))
           nullify(spm2_5)
-!
+
```
nitpick:  What's this space for?

hongli-wang replied:
Deleted a comment line and added a blank line which works as a separation from the next pieces of code.

RussTreadon-NOAA reviewed (commented):
Just a few minor comments.   Do existing regression tests exercise new option(s)?  If not, have we confirmed that the updated code generates the correct results when the new option(s) are activated?

CoryMartin-NOAA commented:
@RussTreadon-NOAA I'm not sure that any of this will end up in operations before our transition to JEDI, but this will serve as a reference for Hongli's science work that will then be soon moved to JEDI for further development. I'm 95% sure these additions are *not* exercised in one of our current regression tests. I don't think it's required (especially since we want to keep the test count low) but something to consider in the future if RRFS-CMAQ or RRFS-Smoke has GSI based DA in ops.

RussTreadon-NOAA commented:
> @RussTreadon-NOAA I'm not sure that any of this will end up in operations before our transition to JEDI, but this will serve as a reference for Hongli's science work that will then be soon moved to JEDI for further development. I'm 95% sure these additions are _not_ exercised in one of our current regression tests. I don't think it's required (especially since we want to keep the test count low) but something to consider in the future if RRFS-CMAQ or RRFS-Smoke has GSI based DA in ops.

Thanks, @CoryMartin-NOAA, for the reply.  This puts things in perspective.

ShunLiu-NOAA commented:
Good points, Cory and Russ.

At the moment, I think at least we should confirm there is no impact on the
current functions (fit-to-obs) of GSI with turning off the option through a
regression test.

Shun

On Mon, Jan 23, 2023 at 4:24 PM Cory Martin ***@***.***>
wrote:

> @RussTreadon-NOAA <https://github.com/RussTreadon-NOAA> I'm not sure that
> any of this will end up in operations before our transition to JEDI, but
> this will serve as a reference for Hongli's science work that will then be
> soon moved to JEDI for further development. I'm 95% sure these additions
> are *not* exercised in one of our current regression tests. I don't think
> it's required (especially since we want to keep the test count low) but
> something to consider in the future if RRFS-CMAQ or RRFS-Smoke has GSI
> based DA in ops.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/GSI/pull/482#issuecomment-1401001105>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACD5GZ63WF6KHVGOPKPUR73WT3ZIZANCNFSM6AAAAAAQ4ZE2IE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hongli-wang commented:
> @RussTreadon-NOAA I'm not sure that any of this will end up in operations before our transition to JEDI, but this will serve as a reference for Hongli's science work that will then be soon moved to JEDI for further development. I'm 95% sure these additions are _not_ exercised in one of our current regression tests. I don't think it's required (especially since we want to keep the test count low) but something to consider in the future if RRFS-CMAQ or RRFS-Smoke has GSI based DA in ops.

@CoryMartin-NOAA @RussTreadon-NOAA I don't know if there is regression test case with a wrfchem background, but there are not any of regression cases are setup with RRFS-CMAQ and RRFS-Smoke background. I agree with Cory that we may have such a test when the new functions are in operation.

hongli-wang commented:
> @ShunLiu-NOAA good point, let me set that up as it has been a while since I've done the process myself.

@ShunLiu-NOAA thanks for pointing this out.  @CoryMartin-NOAA Thanks for taking care the regression test. 

Hongli

RussTreadon-NOAA commented:
> > @ShunLiu-NOAA good point, let me set that up as it has been a while since I've done the process myself.
> 
> @ShunLiu-NOAA thanks for pointing this out. @CoryMartin-NOAA Thanks for taking care the regression test.
> 
> Hongli

I encourage you @hongli-wang to also run the regression tests.  Do you have access to Hera or Orion?

CoryMartin-NOAA commented:
@hongli-wang to echo what Russ said, with these instructions: https://github.com/NOAA-EMC/GSI/wiki/GSI-Ctests-(regression-tests) it is pretty easy to do

hongli-wang commented:
> 

@CoryMartin-NOAA  I will have a try. Thanks.

CoryMartin-NOAA commented:
All 9 tests pass for me on Hera:
```
1/9 Test #7: [=[rrfs_3denvar_glbens]=] ........   Passed  487.37 sec
2/9 Test #9: [=[global_enkf]=] ................   Passed  543.43 sec
3/9 Test #8: [=[netcdf_fv3_regional]=] ........   Passed  724.98 sec
4/9 Test #4: [=[hwrf_nmm_d2]=] ................   Passed  730.10 sec
5/9 Test #5: [=[hwrf_nmm_d3]=] ................   Passed  737.72 sec
6/9 Test #2: [=[global_4dvar]=] ...............   Passed  1503.27 sec
7/9 Test #6: [=[rtma]=] .......................   Passed  1571.90 sec
8/9 Test #1: [=[global_3dvar]=] ...............   Passed  1864.59 sec
9/9 Test #3: [=[global_4denvar]=] .............   Passed  2045.13 sec

100% tests passed, 0 tests failed out of 9

Total Test time (real) = 2045.14 sec
```

hongli-wang commented:
> @hongli-wang to echo what Russ said, with these instructions: https://github.com/NOAA-EMC/GSI/wiki/GSI-Ctests-(regression-tests) it is pretty easy to do

@CoryMartin-NOAA Regarding step #3 running regression test, should I run in the dir GSI/build, which is from my fork?

RussTreadon-NOAA commented:
> > @hongli-wang to echo what Russ said, with these instructions: https://github.com/NOAA-EMC/GSI/wiki/GSI-Ctests-(regression-tests) it is pretty easy to do
> 
> @CoryMartin-NOAA Regarding step #3 running regression test, should I run in the dir GSI/build, which is from my fork?

Yes, @hongli-wang , the ctests are executed from the `build` directory in the working copy of your fork.

RussTreadon-NOAA commented:
@hongli-wang , once requested changes are made and your regression tests (ctest) results are posted I can approve this PR.

hongli-wang committed:
Change variables(iuv,ndynvario3d,ntracerio3d) to i_kind and clear code.

hongli-wang commented:
> @hongli-wang , once requested changes are made and your regression tests (ctest) results are posted I can approve this PR.

@RussTreadon-NOAA I had changed the variables (iuv,ndynvario3d,ntracerio3d) to i_kind in gsi_rfv3io_mod.f90, and deleted unused commented variable (penalty) in intpm2_5.f90. The changes had been pushed back. 
Please let me know if you see other changes are required. Thanks.

RussTreadon-NOAA commented:
Changes look good.  Thank you @hongli-wang .   Would you please post results from your execution of the regression (ctest) tests?

@TingLei-NOAA , you have several unresolved conversations in this PR.   Would you please check these open conversations and mark them as resolved if appropriate.

TingLei-NOAA commented:
@Russ Treadon - NOAA Federal ***@***.***>  Sure, all those
conversations I initiated were resolved. However, I think I don't have the
permission to mark those conversations resolved. I think you (with the
write permissions to EMC GSI ) or @hongli-wang (the author) could do it.
So, either of you do that will be ok to me and thanks !

______________________________

Ting Lei

Physical Scientist, Contractor with Lynker in support of

EMC/NCEP/NWS/NOAA

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

***@***.***

301-683-3624


On Tue, Jan 24, 2023 at 11:18 AM RussTreadon-NOAA ***@***.***>
wrote:

> Changes look good. Thank you @hongli-wang <https://github.com/hongli-wang>
> . Would you please post results from your execution of the regression
> (ctest) tests?
>
> @TingLei-NOAA <https://github.com/TingLei-NOAA> , you have several
> unresolved conversations in this PR. Would you please check these open
> conversations and mark them as resolved if appropriate.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/GSI/pull/482#issuecomment-1402217247>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APEFS7CXTVUSGZF2EE4JO6LWT76G5ANCNFSM6AAAAAAQ4ZE2IE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hongli-wang commented:
> Changes look good. Thank you @hongli-wang . Would you please post results from your execution of the regression (ctest) tests?
> 
> @TingLei-NOAA , you have several unresolved conversations in this PR. Would you please check these open conversations and mark them as resolved if appropriate.

@RussTreadon-NOAA The regression test under my Hera account doesn't work well. 
Yesterday I followed the guidance on regression test, and ran with ctest in the dir build of my fork, it failed. However, I don't think this is a code issue since Cory had a successful test. 

FYI. I do change the regression_var.sh. I am wondering if it is the HPC setup issue. 

Here is the first regression test if you like to have a look. It seems the gsi.x (both develop and mine) cann't handle the namlist file: 
/scratch1/BMC/wrfruc/hwang/smoke/ctest/tmpreg_global_3dvar/global_3dvar_loproc_updat

RussTreadon-NOAA commented:
> @russ Treadon - NOAA Federal ***@***.***> Sure, all those conversations I initiated were resolved. However, I think I don't have the permission to mark those conversations resolved. I think you (with the write permissions to EMC GSI ) or @hongli-wang (the author) could do it. So, either of you do that will be ok to me and thanks !
> […](#)
> ______________________________ Ting Lei Physical Scientist, Contractor with Lynker in support of EMC/NCEP/NWS/NOAA 5830 University Research Ct., Cubicle 2765 College Park, MD 20740 ***@***.*** 301-683-3624
> On Tue, Jan 24, 2023 at 11:18 AM RussTreadon-NOAA ***@***.***> wrote: Changes look good. Thank you @hongli-wang <https://github.com/hongli-wang> . Would you please post results from your execution of the regression (ctest) tests? @TingLei-NOAA <https://github.com/TingLei-NOAA> , you have several unresolved conversations in this PR. Would you please check these open conversations and mark them as resolved if appropriate. — Reply to this email directly, view it on GitHub <[#482 (comment)](https://github.com/NOAA-EMC/GSI/pull/482#issuecomment-1402217247)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/APEFS7CXTVUSGZF2EE4JO6LWT76G5ANCNFSM6AAAAAAQ4ZE2IE> . You are receiving this because you were mentioned.Message ID: ***@***.***>

@TingLei-NOAA , did you try resolving the conversations by clicking on `Resolve conversation`?

RussTreadon-NOAA commented:
> 

@hongli-wang , if you want me to examine this failure I need to know the path to the build directory from which you executed `ctest`.

hongli-wang commented:
> @hongli-wang , if you want me to examine this failure I need to know the path to the build directory from which you executed `ctest`.

@RussTreadon-NOAA 
My gsi: 
/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/build
EMC develop:
/scratch1/BMC/wrfruc/hwang/smoke/ctest/develop

Thanks for looking into this.

hongli-wang commented:
> @russ Treadon - NOAA Federal ***@***.***> Sure, all those conversations I initiated were resolved. However, I think I don't have the permission to mark those conversations resolved. I think you (with the write permissions to EMC GSI ) or @hongli-wang (the author) could do it. So, either of you do that will be ok to me and thanks !
> […](#)

@TingLei-NOAA @RussTreadon-NOAA 
I went though conversations and had makes a few as resolved where I have the permission to change their status. Thank you all for your time on this PR. Hongli

RussTreadon-NOAA commented:
@hongli-wang , it looks like you did not include `--recursive` when you executed `git clone` to populate `/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI`.   Directory `/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix` is empty.   

Log file `/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/regression/global_3dvar_loproc_updat.out` contains numerous `No such file` messages.
```
+ /bin/cp /scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_anavinfo.l127.txt ./anavinfo
/bin/cp: cannot stat ?/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_anavinfo.l127.txt?: No such file or directory
+ /bin/cp /scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/Big_Endian/global_berror.l127y98.f77 ./berror_stats
/bin/cp: cannot stat ?/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/Big_Endian/global_berror.l127y98.f77?: No such file or directory
+ /bin/cp /scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_hybens_info.l127.txt ./hybens_info
/bin/cp: cannot stat ?/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_hybens_info.l127.txt?: No such file or directory
```

@hu5970 , step 1 of `GSI Ctests (regression tests)` needs `--recursive` on the `git clone`.

TingLei-NOAA commented:
@hongli-wang Thanks for marking them resolved. 
@RussTreadon-NOAA   no "resolved " button appear to me , i guess, for the permission issue I mentioned previously.

RussTreadon-NOAA commented:
> @hongli-wang Thanks for marking them resolved. @RussTreadon-NOAA no "resolved " button appear to me , i guess, for the permission issue I mentioned previously.

Thank you @TingLei-NOAA for resolving the conversations you initiated.  Please do so in future reviews.  The review team can not process a PR if conversations remain open.

hongli-wang commented:
> @hongli-wang , it looks like you did not include `--recursive` when you executed `git clone` to populate `/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI`. Directory `/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix` is empty.
> 
> Log file `/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/regression/global_3dvar_loproc_updat.out` contains numerous `No such file` messages.
> 
> ```
> + /bin/cp /scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_anavinfo.l127.txt ./anavinfo
> /bin/cp: cannot stat ?/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_anavinfo.l127.txt?: No such file or directory
> + /bin/cp /scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/Big_Endian/global_berror.l127y98.f77 ./berror_stats
> /bin/cp: cannot stat ?/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/Big_Endian/global_berror.l127y98.f77?: No such file or directory
> + /bin/cp /scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_hybens_info.l127.txt ./hybens_info
> /bin/cp: cannot stat ?/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_hybens_info.l127.txt?: No such file or directory
> ```
> 
> @hu5970 , step 1 of `GSI Ctests (regression tests)` needs `--recursive` on the `git clone`.

@RussTreadon-NOAA You are correct. I did not include `--recursive` when I checked my GSI yesterday. I will have another try. thanks.

TingLei-daprediction commented:
> > @hongli-wang Thanks for marking them resolved. @RussTreadon-NOAA no "resolved " button appear to me , i guess, for the permission issue I mentioned previously.
> 
> Thank you @TingLei-NOAA for resolving the conversations you initiated. Please do so in future reviews. The review team can not process a PR if conversations remain open.

@RussTreadon-NOAA  As I mentioned , those conversations could only be "resolved" by persons with write permission and the author, and hence, I have no "right" to "resolve " conversations.  (this time , Hongli resolved those conversations. ).   Hope this explanation make  sense to you.

RussTreadon-NOAA commented:
> > > @hongli-wang Thanks for marking them resolved. @RussTreadon-NOAA no "resolved " button appear to me , i guess, for the permission issue I mentioned previously.
> > 
> > 
> > Thank you @TingLei-NOAA for resolving the conversations you initiated. Please do so in future reviews. The review team can not process a PR if conversations remain open.
> 
> @RussTreadon-NOAA As I mentioned , those conversations could only be "resolved" by persons with write permission and the author, and hence, I have no "right" to "resolve " conversations. (this time , Hongli resolved those conversations. ). Hope this explanation make sense to you.

@TingLei-NOAA .   I searched github documentation and [see](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/about-pull-request-reviews#resolving-conversations) what you are saying.   It's always good to learn new things.   Thanks.

TingLei-NOAA commented:
> 

@RussTreadon-NOAA  Thanks for the confirmation. Sure, that is also what I learned not long ago!. Thanks.

hongli-wang commented:
> > @hongli-wang , it looks like you did not include `--recursive` when you executed `git clone` to populate `/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI`. Directory `/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix` is empty.
> > Log file `/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/regression/global_3dvar_loproc_updat.out` contains numerous `No such file` messages.
> > ```
> > + /bin/cp /scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_anavinfo.l127.txt ./anavinfo
> > /bin/cp: cannot stat ?/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_anavinfo.l127.txt?: No such file or directory
> > + /bin/cp /scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/Big_Endian/global_berror.l127y98.f77 ./berror_stats
> > /bin/cp: cannot stat ?/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/Big_Endian/global_berror.l127y98.f77?: No such file or directory
> > + /bin/cp /scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_hybens_info.l127.txt ./hybens_info
> > /bin/cp: cannot stat ?/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_hybens_info.l127.txt?: No such file or directory
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @hu5970 , step 1 of `GSI Ctests (regression tests)` needs `--recursive` on the `git clone`.
> 
> @RussTreadon-NOAA You are correct. I did not include `--recursive` when I checked my GSI yesterday. I will have another try. thanks.

@RussTreadon-NOAA @CoryMartin-NOAA My regression test is pending and no progress. Just wondering if any of you can run a regression test. Thanks.

RussTreadon-NOAA commented:
> > > @hongli-wang , it looks like you did not include `--recursive` when you executed `git clone` to populate `/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI`. Directory `/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix` is empty.
> > > Log file `/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/regression/global_3dvar_loproc_updat.out` contains numerous `No such file` messages.
> > > ```
> > > + /bin/cp /scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_anavinfo.l127.txt ./anavinfo
> > > /bin/cp: cannot stat ?/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_anavinfo.l127.txt?: No such file or directory
> > > + /bin/cp /scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/Big_Endian/global_berror.l127y98.f77 ./berror_stats
> > > /bin/cp: cannot stat ?/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/Big_Endian/global_berror.l127y98.f77?: No such file or directory
> > > + /bin/cp /scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_hybens_info.l127.txt ./hybens_info
> > > /bin/cp: cannot stat ?/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/fix/global_hybens_info.l127.txt?: No such file or directory
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > @hu5970 , step 1 of `GSI Ctests (regression tests)` needs `--recursive` on the `git clone`.
> > 
> > 
> > @RussTreadon-NOAA You are correct. I did not include `--recursive` when I checked my GSI yesterday. I will have another try. thanks.
> 
> @RussTreadon-NOAA @CoryMartin-NOAA My regression test is pending and no progress. Just wondering if any of you can run a regression test. Thanks.

@hongli-wang , let me try later tonight.

RussTreadon-NOAA commented:
ctests work for me
```
Hera(hfe12):/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr482/build$ ctest -j 9
Test project /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr482/build
    Start 1: global_3dvar
    Start 2: global_4dvar
    Start 3: global_4denvar
    Start 4: hwrf_nmm_d2
    Start 5: hwrf_nmm_d3
    Start 6: rtma
    Start 7: rrfs_3denvar_glbens
    Start 8: netcdf_fv3_regional
    Start 9: global_enkf
1/9 Test #7: rrfs_3denvar_glbens ..............   Passed  485.53 sec
2/9 Test #8: netcdf_fv3_regional ..............   Passed  542.94 sec
3/9 Test #9: global_enkf ......................   Passed  601.69 sec
4/9 Test #5: hwrf_nmm_d3 ......................   Passed  674.07 sec
5/9 Test #4: hwrf_nmm_d2 ......................   Passed  787.56 sec
6/9 Test #2: global_4dvar .....................   Passed  1321.79 sec
7/9 Test #6: rtma .............................   Passed  1449.58 sec
8/9 Test #3: global_4denvar ...................   Passed  1803.40 sec
9/9 Test #1: global_3dvar .....................   Passed  1922.81 sec

100% tests passed, 0 tests failed out of 9

Total Test time (real) = 1922.81 sec
Hera(hfe12):/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr482/build$ 
```

My `pr482` directory is the current head of your `feature/RRFS-CMAQ`
```
Hera(hfe12):/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr482/build$ git branch
* feature/RRFS-CMAQ
  master
Hera(hfe12):/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr482/build$ git log --oneline |head -3
08b212f4c Change variables(iuv,ndynvario3d,ntracerio3d) to i_kind and clear code.
b6ab52aa5 update submodule to resolve change
7dbfb184c Merge remote-tracking branch 'upstream/develop' into feature/RRFS-CMAQ
```

RussTreadon-NOAA commented:
@TingLei-daprediction , @TingLei-NOAA would you please add your review to this PR.   

@hongli-wang , any luck with the regression tests?

hongli-wang commented:
> @TingLei-daprediction , @TingLei-NOAA would you please add your review to this PR.
> 
> @hongli-wang , any luck with the regression tests?

@RussTreadon-NOAA The test of global_enkf failed. Others passed. 
  Start 1: [=[global_3dvar]=]
    Start 2: [=[global_4dvar]=]
    Start 3: [=[global_4denvar]=]
    Start 4: [=[hwrf_nmm_d2]=]
    Start 5: [=[hwrf_nmm_d3]=]
    Start 6: [=[rtma]=]
    Start 7: [=[rrfs_3denvar_glbens]=]
    Start 8: [=[netcdf_fv3_regional]=]
    Start 9: [=[global_enkf]=]
1/9 Test #7: [=[rrfs_3denvar_glbens]=] ........   Passed  1327.02 sec
2/9 Test #9: [=[global_enkf]=] ................***Failed  1443.16 sec
3/9 Test #8: [=[netcdf_fv3_regional]=] ........   Passed  1624.90 sec
4/9 Test #4: [=[hwrf_nmm_d2]=] ................   Passed  1629.21 sec
5/9 Test #5: [=[hwrf_nmm_d3]=] ................   Passed  1636.73 sec

6/9 Test #2: [=[global_4dvar]=] ...............   Passed  2044.37 sec
7/9 Test #3: [=[global_4denvar]=] .............   Passed  2209.98 sec
8/9 Test #6: [=[rtma]=] .......................   Passed  2232.52 sec
9/9 Test #1: [=[global_3dvar]=] ...............   Passed  2510.73 sec

89% tests passed, 1 tests failed out of 9

Total Test time (real) = 2510.74 sec

The following tests FAILED:
	  9 - [=[global_enkf]=] (Failed)

My build on Hera: 
/scratch1/BMC/wrfruc/hwang/smoke/ctest/GSI/build/

Please feel free to have a look. Thanks again.

RussTreadon-NOAA commented:
@hongli-wang , a check of file `global_enkf_regression_results.txt`  in `/scratch1/BMC/wrfruc/hwang/smoke/ctest/tmpreg_global_enkf/tmpreg_global_enkf/global_enkf_loproc_updat_vs_global_enkf_loproc_contrl/` shows that the failure is due to a run time check.

```
The runtime for global_enkf_hiproc_updat is 64.759394 seconds.  This has exceeded maximum allowable threshold time of 64.704082 seconds,
resulting in Failure timethresh2 of the regression test.
```

The difference is 0.05 second.  This is a non-fatal fail.

Given @CoryMartin-NOAA and @TingLei-NOAA  approvals, I'll merge this PR into `develop`.

CoryMartin-NOAA commented:
Thanks @RussTreadon-NOAA for helping finish this up, I wanted a second set of eyes given that I contributed through the merging of develop + modifying the `fix/` path.

hongli-wang commented:
> Thanks @RussTreadon-NOAA for helping finish this up, I wanted a second set of eyes given that I contributed through the merging of develop + modifying the `fix/` path.

@CoryMartin-NOAA @TingLei-NOAA @RussTreadon-NOAA @ShunLiu-NOAA @hu5970 
Thanks Cory for merging of develop and updating fix in my fork. I am not familiar with that part, that saves me tons of time! Thanks Cory and Ting for reviews since the DA work for RRFS-CMAQ and now RRFS-smoke. 

Thank you ALL for your time and contributions!  Appreciated! 

Hongli
